test/filters/test_crop_hull.cpp

```diff
@@ -0,0 +1,378 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2012-, Open Perception, Inc.
+ *
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder(s) nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * $Id: test_filters.cpp 7683 2012-10-23 02:49:03Z rusu $
+ *
+ */
+
+#include <pcl/test/gtest.h>
+#include <pcl/pcl_tests.h>
+
+#include <random>
+
+#include <pcl/point_types.h>
+#include <pcl/surface/convex_hull.h>
+#include <pcl/common/common.h>
+#include <pcl/filters/crop_hull.h>
+
+
+namespace
+{
+  template <class PointGenerator1, class PointGenerator2>
+  void complexTest(
```

Comment:
Please refactor the tests into:
1. Test Fixture (cloud generator)
2. Separate equivalence function/class (currently `CheckEquals`)
3. Separate test composer function (loop over parameters of `cropHullFilter` and test using the equivalence class).
4. Separate composer to test (currently testing using upper, lower and empty indices)
5. Actual implementation of tests setup using the fixture (1), using the outer composer (4) which calls the inner composer (3) which checks using equivalence (2)

I don't think parametrized test in gtest will be a good fit here, but maybe I'm wrong.

Reply:
Actually, I don't fully understand this list. However, I tried to follow list as could.
1. Extract cloud generator as individual function. Lower, upper etc. indices fused in vector of `pair<Indices, PointCloud>`
1. Separate class `CheckEquals` as individual class `Checker`. My fault in name. This class don't check equivalence, but call test suites.
1. Change template types to `std::function<pcl::PointXYZ()>`

Reply:
I've added more comments explaining the reasoning locally. Let's work iteratively on these suggestions. You changes were in the correct direction.